daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/svalue/Hasher.scala

```diff
@@ -0,0 +1,101 @@
+  private final case class Mix(data: Int) extends Command
+  // mix way the n first elements of the stack in a single block
+  private final case class Ordered(n: Int) extends Command
+  // mix in a symmetric way the n first elements of the stack in a single block
+  private final case class Unordered(n: Int) extends Command
+  // compute the hash of a value
+  private final case class Value(v: SValue) extends Command
+
+  def hash(v: SValue): Int =
+    loop(List(Value(v)))
+
+  private def pushOrderedValues(values: Iterator[SValue], cmds: List[Command]) =
+    ((Ordered(values.size) :: cmds) /: values) { case (acc, v) => Value(v) :: acc }
+
+  @tailrec
+  private def loop(cmds: List[Command], stack: List[Int] = List.empty): Int =
+    cmds match {
+      case cmd :: cmdsRest =>
+        cmd match {
+          case Value(v) =>
+            v match {
+              case _: SPAP | SToken | _: STNat =>
+                throw NonHashableSValue(v)
+              case SUnit =>
+                loop(cmdsRest, 0 :: stack)
+              case SBool(b) =>
+                loop(cmdsRest, b.hashCode() :: stack)
+              case SInt64(i) =>
+                loop(cmdsRest, i.toInt :: stack)
+              case SNumeric(n) =>
+                loop(cmdsRest, n.hashCode() :: stack)
+              case SText(s) =>
+                loop(cmdsRest, s.hashCode :: stack)
+              case SDate(d) =>
+                loop(cmdsRest, d.hashCode() :: stack)
+              case STimestamp(t) =>
+                loop(cmdsRest, t.hashCode() :: stack)
+              case SParty(p) =>
+                loop(cmdsRest, p.hashCode :: stack)
+              case SContractId(cid) =>
+                loop(cmdsRest, cid.hashCode :: stack)
+              case STypeRep(t) =>
+                loop(cmdsRest, t.hashCode() :: stack)
+              case SEnum(_, constructor) =>
+                loop(cmdsRest, constructor.hashCode :: stack)
+              case SRecord(_, _, values) =>
+                loop(pushOrderedValues(values.iterator().asScala, cmdsRest), stack)
+              case SVariant(_, variant, value) =>
+                loop(Value(value) :: Mix(variant.hashCode) :: cmdsRest, stack)
+              case STuple(_, values) =>
+                loop(pushOrderedValues(values.iterator().asScala, cmdsRest), stack)
+              case SOptional(opt) =>
+                loop(pushOrderedValues(opt.iterator, cmdsRest), stack)
+              case SList(values) =>
+                loop(pushOrderedValues(values.iterator, cmdsRest), stack)
+              case SMap(value) =>
+                val newCmds = ((Unordered(value.size) :: cmdsRest) /: value) {
+                  case (acc, (k, v)) => Value(v) :: Mix(k.hashCode) :: acc
+                }
+                loop(newCmds, stack)
+              case SGenMap(_, values) =>
+                val newCmds = ((Unordered(values.size) :: cmdsRest) /: values) {
+                  case (acc, (k, (_, v))) => Value(v) :: Mix(k.hashCode) :: acc
```

Comment:
In this case `k.hashCode` will result in a recursive call to Hasher.hash.
Is this the expected behavior?

The recursive call can be removed by adding a second `Mix` command that takes both arguments from the stack.

Reply:
The `hashCode` field of a `SGenMap.Key` is a `val`. It is compute once when the object is construct. There is no recursive call here.

Reply:
^ that makes sense, thanks